### ISSUE TITLE
[Feature Request]: Non-finite gradient can be accumulated during the gradient accumulation

### ISSUE THREAD

**Issue by jaeyeonkim99** [speechbrain/speechbrain#1936]
### 🚀The feature

![image](https://user-images.githubusercontent.com/84221374/231334266-18acba61-d621-46a5-abd7-ca52b7a59766.png)

In the current `fit_batch()` function of the `Brain` class, `check_graidents()` is called only on the `should_step`(the step for optimization after the gradient is accumulated for the gradient accumulation steps).
However, if the loss is not finite on non-should steps, 'check_gradients()' isn't called so the non-finite gradients are accumulated. Then if the loss is finite on the following 'should_step', the 'optimzation.step()' will run with accumulated non-finite gradients. 

### Solution outline

Now, the 'check_graidents()' performs 1. check if the loss is finite 2. gradient clipping.
I think these two functionalities should be separated. Checking if the loss is finite should be done for every step(including both should and non-should steps) before `backward()` is called. Then, gradient clipping should be called for the `should_step` only, 

It should be something like this. 
![image](https://user-images.githubusercontent.com/84221374/231342602-cea933f1-2474-4096-af35-623478502237.png)

**Comment by Adel-Moumen** [comment]
Hi @jaeyeonkim99,

Thanks for spotting this. 

Could you please open a PR on that? 

Many thanks

**Comment by jaeyeonkim99** [comment]
No problem.
I will open a PR on this soon.

**Comment by jaeyeonkim99** [comment]
I was busy, and just open a PR on this.

**Comment by jaeyeonkim99** [comment]
I've just open the PR on this issue.

**Comment by Adel-Moumen** [comment]
Solved in #2018. Thanks gain for the PR! :-)
